src/content/plugins/internal-plugins.md

```diff
@@ -163,104 +187,142 @@ if (a) {
 }
 ```
 
-Becausse `a` will evaluate as `false` in the `if (a)` condition and `foo()` call will never happen, the whole import will get pruned thanks to `ConstPlugin` run.
+Because `a` will evaluate as `false` in the `if (a)` condition and `foo()` call will never happen, the whole import will get pruned thanks to `ConstPlugin` run.
+
```

Comment:
This doesn't sound that well:

>  the whole import will get pruned thanks to `ConstPlugin` run

..will get pruned by the `ConstPlugin`

.

Reply:
i am not quite sure, because import pruning maybe happens via uglify, but if not const plugin that wouldnt be possible

Reply:
thats what i was trying to say by "thanks to ..."

Reply:
Probably yes, I would assume so.
`UglifyJS` is injected, when `--mode production` (treeshaking).

I guess `webpack` recognizes then, after the `ConstPlugin` has removed dead branches that there is an unused `import` and removes it, right?

Reply:
yeah so there are two options and would be nice if we can confirm:

a) by the time code gets to uglify it contains if (false) {...} produced by const plugin, so at this point uglify eliminates both import and if (false) {} branch.
b) by the time code gets to uglify it only contains unused import that uglify removes, because if (false) {} was evaluated and removed by const plugin

My guess is that option a) is correct and const plugin only does evaluate conditionals and tries to put false/true where possible, after that uglify removes both branches and imports. But as i said would be nice to turn uglify off and try

Reply:
There is an open [issue](https://github.com/webpack/webpack/issues/7519) regarding ConstPlugin

Reply:
Yes but it doesnt really say that anything except ternary clearing isnt working, i still think its done by uglify. @sokra, @ooflorent can you clarify please

Reply:
There are multiple optimizations in production mode regarding dead branches:

- The ones performed by UglifyJS
- The ones performed by webpack

webpack will try to evaluate conditional statements. If it succeeds then the dead branch is removed. webpack can't do constant folding unless the compiler knows it.

```js
const FOO = 1
if (FOO === 0) { /* dead */ }
```

In the above example, webpack is unable to prune the branch, but Uglify does. However, if `FOO` is defined using `DefinePlugin`, webpack will succeed.

```js
if (0 === 1) { /* dead */ }
```

Here webpack is able to reduce the comparison and will remove the branch.

> except ternary clearing isnt working

It also works for ternaries but not with _sloppy_ comparisons (see the issue mentioned by Legends).

Reply:
Pretty much what we though except for the define plugin detail, thank you very much @ooflorent for stepping by. i will update this one